Added Aspect Ratio Calculator Panel

roamindev opened the pull request:
**Summary**

Added aspect ratio calculator panel. Allows user to enter a starting width and height and calculate new width or height based on the aspect ratio of the first set of values. Can also apply the calculated values to the current tab's (txt2img or img2img) width and height sliders. Inspired by the aspect ratio calculator website by Andrew Hedges (https://andrew.hedges.name/experiments/aspect_ratio/)

**Details**
- Show/hide calculator panel using new calculator button on Aspect Ratio Buttons row
- Swap width and height values
- Multiple ways to set base width and height values:
  - Manually entered by user
  - Pull values from current tab's width and height sliders
  - Pull values from the image in the current img2img tab's image input component
- Apply new calculated width and height values to the current tab's width and height sliders

**Reason and Notes**

I noticed I use various web-based AR calculators like the one mentioned in the summary a lot in my workflows and I got tired of the tediously slow process of getting the dimensions of a source image, pulling up a calculator website to calculate new values, and then manually inputting those values back into the SD web ui. I've found it's most useful in img2img mode when using a source image that didn't come from SD (i.e. unknown dimensions), or when using ControlNet with the same in either mode.

I thought about coding a separate extension, but it seemed more appropriate to build it into this one (that I was already using), and I think it's a good fit. With that said, this is my first try at writing code in Python. I tried to follow the flow/structure of the extension's code as best I could. Please let me know if there's anything that needs to be updated or done differently. It's also my first attempt at submitting a pull request on GitHub, so hopefully I've done that correctly as well.

Lastly, no hard feelings if you're not interested in adding it in. I've just found it extremely useful already and figured I should share.

**Screenshot**

![image](https://user-images.githubusercontent.com/121050401/229013783-1b9cf303-0e75-44f3-9b3f-20a493b90623.png)

roamindev committed:
Added aspect ratio calculator

Added aspect ratio calculator panel inspired by the aspect ration calculator website by Andrew Hedges (https://andrew.hedges.name/experiments/aspect_ratio/)

alemelis commented on scripts/sd-webui-ar.py:
```diff
@@ -122,6 +127,47 @@ def write_resolutions_file(filename):
 
 
+## Functions for calculator panel
+def gcd(a, b):
```
love the recursive approach but perhaps we can use the standard library for this

```python
from math import gcd
```

roamindev replied:
Absolutely, makes more sense to do that. Total Python noob here, but I probably should've guessed there was a library for it.

(obligatory xkcd: https://xkcd.com/353/) :)

alemelis commented on scripts/sd-webui-ar.py:
```diff
@@ -122,6 +127,47 @@ def write_resolutions_file(filename):
+    if n == d:
+        return '1:1'
+
+    if n < d:
```
here it gets shorter as 

```python
if n < d:
    n, d = d, n
```

roamindev replied:
Much cleaner. I like this.

alemelis commented on scripts/sd-webui-ar.py:
```diff
@@ -122,6 +127,47 @@ def write_resolutions_file(filename):
+        n = d
+        d = temp
+
+    div = gcd(int(n), int(d))
```
do we expect `n` and `d` to be not integers?
In that case, it's possibly better to cast them once at the beginning

```python
n, d = list(map(int, (n, d)))
```

roamindev replied:
This is a byproduct of my port from JS to Python. The original JS code had `(+numerator, +denominator)` to ensure they were numbers and I just chose `int()` over `float()`. It appears they come from the gradio `Number` components as `float` values, but now that I've switched to `math.gcd`, it wants `int` values, so I think it is needed now (albeit, with the code you provided, instead).

alemelis commented on scripts/sd-webui-ar.py:
```diff
@@ -122,6 +127,47 @@ def write_resolutions_file(filename):
+
+    div = gcd(int(n), int(d))
+
+    if 'temp' not in locals():
```
can you explain better what is the code in 148-158 achieving?
not sure that looking in `locals()` is best practice

roamindev replied:
For lines 148-153: The original JS code checked to see if `temp` was undefined to know whether or not we'd swapped the numerator and denominator to calculate gcd so they could be swapped back if needed. I wasn't sure how to do this in Python and checking `locals()` is what I came up with after a bit of searching. With the change you recommended to use `n, d = d, n` to swap the values, `temp` no longer exists at all, so I think I'll need a swapped flag instead. i.e....

```Python
    swapped = False
    if n < d:
        n, d = d, n
        swapped = True

    div = gcd(n, d)

    if swapped:
        w = int(n) // div
        h = int(d) // div
    else:
        w = int(d) // div
        h = int(n) // div
```
Let me know if there's a better way to do this in Python.

Lines 155-157 just rename a ratio of 8:5 to 16:10. 16:10 is the AR for WUXGA (1920x1200), sometimes called a "taller" version of 1080p (even though it's not quite 1080p). An extremely rare case, for sure.

alemelis replied:
I guess you can rewrite the entire function as 

```python
def get_reduced_ratio(n, d):
    n, d = list(map(int, (n, d)))

    if n == d:
        return '1:1'

    if n<d:
        n, d = d, n
        div = gcd(n, d)
        w = n // div
        h = d // div
    else:
        div = gcd(n, d)
        w = d // div
        h = n // div

    if w == 8 and h == 5:
        w = 16
        h = 10

    return f'{w}:{h}'
```

so that `locals` is not called and there's no need for a temp variable (yes, `gcd` is called twice but we can live with ti)

roamindev replied:
This works great and allows it to be even simpler. If I just swap the order they're sent to `gcd` in, then the output to `w` and `h` is the same for both cases.

```Python
    if n<d:
        div = gcd (d, n)
    else:
        div = gcd(n, d)

    w = int(n) // div
    h = int(d) // div
```

alemelis commented on scripts/sd-webui-ar.py:
```diff
@@ -212,3 +380,15 @@ def after_component(self, component, **kwargs):
             self.i2i_h = component
+
+        if kwargs.get("elem_id") == "img2img_image":
+            self.image = []
```
is `self.image` first defined here?
perhaps better to place it on line 383

roamindev replied:
I had originally tried it there, but when it's outside of that initial `if` statement, `self.image[i]` comes back empty when referenced. I wasn't sure about what fired when in that `after_component()` section to know if I was missing something, but putting the initial definition inside the first img2img tab `if` statement was the only way I could get it to work properly.

alemelis replied:
I see, in that case you can create do it in one line

```python
        if kwargs.get("elem_id") == "img2img_image":
            self.image = [component]
```

roamindev replied:
Perfect! Makes sense once I see it. Have to admit that section gave me more headache than it probably should've. :)

alemelis reviewed (commented):
Hello! thanks a lot for this contribution, I like it a lot!

I'm happy to add this to the extension and the code looks good already. There are just few things I'd like to change/clarify

roamindev commented:
> Hello! thanks a lot for this contribution, I like it a lot!
> 
> I'm happy to add this to the extension and the code looks good already. There are just few things I'd like to change/clarify

Awesome! I'm happy to make these changes and really appreciate the notes and guidance. I have a busy weekend ahead but will try to make some time to work on these.

roamindev committed:
Make recommended review changes

Made recommended review changes except where noted in review comments.

roamindev commented:
@alemelis Had some free time for a few so just pushed a new commit with all the recommended changes except for the move of `self.image = []` as noted above (let me know if I'm missing something there). I'm also not sure if there's a better way to handle the replacement of looking in `locals()` for `temp` than using a `swapped` flag variable like I came up with.

Thanks for your help and guidance with all of this! :)

roamindev committed:
Update README.md with calculator instructions

Added section for calculate instructions with animated GIF showing basic usage.

roamindev commented:
Added basic calculator instructions to README.md

roamindev committed:
Update README.md Fixed Typo

Correcting typo in new calculator instructions - wording for the two source dimensions buttons was inconsistent.

alemelis commented on scripts/sd-webui-ar.py:
```diff
@@ -201,6 +249,123 @@ def ui(self, is_img2img):
+                                )
+
+                # Update aspect ratio display on change
+                arc_width1.change(lambda w, h: ("""Aspect Ratio: **{}**""".format(get_reduced_ratio(w,h))), inputs=[arc_width1, arc_height1], outputs=[arc_ar_display])
```
you shouldn't need `"""` here and you can also use [f-strings](https://peps.python.org/pep-0498/) instead of `format`

```python
f"Aspect Ratio: **{get_reduced_ratio(w,h)}**"
```

alemelis replied:
same on line 333

roamindev replied:
Thanks, updated both of these and swapped out the 3x double-quotes elsewhere as well.

alemelis commented:
<img width="368" alt="Screenshot 2023-04-02 at 21 20 30" src="https://user-images.githubusercontent.com/4661737/229377060-500471ff-690f-4663-88e3-e2f88d4ee93e.png">

this is how the calculator icon looks on my browser (Safari). Would it be possible to switch the icon to a text label (say, `calc` or something) in case the browser cannot render the image?

roamindev committed:
Simplify ratio calculation and other changes

Simplified ratio calculation, swapped calculator button icon for text to ensure browser compatibility, updated button styles, simplified self.image declaration.

roamindev committed:
Updated README.md with "Calc" button change

Corrected wording and replaced animated screenshot for new "Calc" button

roamindev commented:
@alemelis Changes made. Also updated the wording and screenshot in README.md calculator section to reflect changes to the `Calc` button.

alemelis reviewed (approved):
LGTM, thanks for the contribution!

roamindev committed:
Add JS based button titles

Add button titles on UI load. Use comments in aspect_ratios.txt and resolutions.txt to for user configured buttons, and hard-coded titles in sd-webui-ar.js for built-in buttons (i.e. for calculator).

roamindev commented:
Pushed commit to address issue #5.

#### Summary
Adds titles (tooltips) to buttons on UI load.

#### Details
Writes `button_titles.js` file each time the UI starts that holds user configured titles read from `aspect_ratios.txt` and `resolutions.txt`. Those get combined with static titles for built-in buttons defined in `sd_webui_ar.js` and then used to set button titles by the `OnUiUpdate` function located in `sd_webui_ar.js`.

#### Notes
Must ship with `button_titles.js` file (can be empty). This file must already exist on first run or the user-configured titles won't get set properly until the second run. `button_titles.js` file gets (over)written on each UI (re)start with whatever is read from `aspect_ratios.txt` and `resolutions.txt`.

#### Screenshot
![image](https://user-images.githubusercontent.com/121050401/229841553-88587494-b0a5-40ec-937c-c5e533c15311.png)

alemelis commented:
@roamindev nice tooltips!

however when I try this branch, only the one on the `Calc` button shows up

also, after this change, the `Calc` button doesn't open the calculator anymore (the other buttons work fine)

roamindev commented:
> @roamindev nice tooltips!
> 
> however when I try this branch, only the one on the `Calc` button shows up
> 
> also, after this change, the `Calc` button doesn't open the calculator anymore (the other buttons work fine)

Okay, I probably messed something up trying to bring in the changes from my tooltips branch. I'll try to look into it this evening and figure out what went wrong.

roamindev commented:
@alemelis I'm not sure what might be wrong. Everything seems to be working fine for me and I can't reproduce the issues you mentioned. I even completely removed the extension and re-cloned the repo to make sure my local install matched what'd been pushed to GitHub.

Are you seeing any sort of errors in the browser's JS console?

![image](https://user-images.githubusercontent.com/121050401/230515919-d889e8a8-1756-4323-8f1b-d11ca883e7b9.png)

alemelis commented:
> @alemelis I'm not sure what might be wrong. Everything seems to be working fine for me and I can't reproduce the issues you mentioned. I even completely removed the extension and re-cloned the repo to make sure my local install matched what'd been pushed to GitHub.
> 
> Are you seeing any sort of errors in the browser's JS console?
> 
> ![image](https://user-images.githubusercontent.com/121050401/230515919-d889e8a8-1756-4323-8f1b-d11ca883e7b9.png)

I'm getting this when launching WebUI

```
Error calling: /Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar/scripts/sd-webui-ar.py/ui
Traceback (most recent call last):
  File "/Users/ale/repos/stable-diffusion-webui/modules/scripts.py", line 243, in wrap_call
    res = func(*args, **kwargs)
  File "/Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar/scripts/sd-webui-ar.py", line 262, in ui
    write_js_titles_file(button_titles)
  File "/Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar/scripts/sd-webui-ar.py", line 141, in write_js_titles_file
    with open(filename, "w", encoding="utf-8") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar\\javascript\\/button_titles.js'

Error calling: /Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar/scripts/sd-webui-ar.py/ui
Traceback (most recent call last):
  File "/Users/ale/repos/stable-diffusion-webui/modules/scripts.py", line 243, in wrap_call
    res = func(*args, **kwargs)
  File "/Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar/scripts/sd-webui-ar.py", line 262, in ui
    write_js_titles_file(button_titles)
  File "/Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar/scripts/sd-webui-ar.py", line 141, in write_js_titles_file
    with open(filename, "w", encoding="utf-8") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/ale/repos/stable-diffusion-webui/extensions/sd-webui-ar\\javascript\\/button_titles.js'

[tag-editor] Settings has been read from config.json
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
```

roamindev committed:
Fix JS File Path Variable

Corrected Path variable for button_titles.js file to proper formatting/usage.

roamindev commented:
> I'm getting this when launching WebUI

@alemelis Latest commit should fix the path error.

The branch should include an empty `button_titles.js` file in `sd-webui-ar/javascript/` that only has a comment and an empty `ar_button_titles` object. As long as that file exists then all of the button titles, both preset and user-specified, should load properly on the first run.

If, for some reason, the `button_titles.js` file does **not** exist, the script will create it but then only the preset titles for the `Calc` button and the buttons in the calculator panel will work on first run, and the user-specified titles for the aspect ratio and resolution buttons would load on the second run (and thereafter).

alemelis reviewed (approved):
<img width="654" alt="Screenshot 2023-04-12 at 10 09 16" src="https://user-images.githubusercontent.com/4661737/231410964-2f9c3926-9b07-4cb7-ad38-afae04941d4a.png">

looking good! thanks for this contribution 🚀

roamindev commented:
> looking good! thanks for this contribution 🚀

Absolutely, my pleasure! I hope others find it as useful as I have.
